File: [0123][Hard][Best_Time_to_Buy_and_Sell_Stock_III]/Best_Time_to_Buy_and_Sell_Stock_III.java
// solution 1, dp with general solution.
// leetcode time     cost : 4 ms
// leetcode memory   cost : 40.6 MB 
// Time  Complexity: O(kN)
// Space Complexity: O(k+N)

import java.util.Arrays;
class Solution {
    public int maxProfit(int[] prices) {
        int max_k = 2;
        int n = prices.length;
        if (n == 0) 
            return 0;
        int dp[][][] = new int[n][max_k + 1][2];
        for (int i = 0; i < n; i++) {
            for (int k = max_k; k >= 1; k--) {
                if (i == 0) { 
                    dp[i][k][0]  = Math.max(0, Integer.MIN_VALUE + prices[i]) ;     //dp[i][k][0]  = 0;
                    dp[i][k][1]  = Math.max(Integer.MIN_VALUE, 0 - prices[i]) ;     //dp[i][k][0]  = - prices[0];
                    continue;
                }
                dp[i][k][0] = Math.max(dp[i-1][k][0], dp[i-1][k][1] + prices[i]);
                dp[i][k][1] = Math.max(dp[i-1][k][1], dp[i-1][k-1][0] - prices[i]);
            }
        }
        // enumerate n × max_k × 2 status.
        return dp[n - 1][max_k][0];
    }
}

public class Best_Time_to_Buy_and_Sell_Stock_III {
    public static void main(String args[]) {
        int prices[] = {1,3,5,0,0,3,1,4}; // #expect is 8
        int prices_2[] = {3,2,6,5,0,3}; // #expect is 7
        Solution Solution_obj = new Solution();
        int result = Solution_obj.maxProfit(prices);
        int result2 = Solution_obj.maxProfit(prices_2);
        System.out.println("In java code, 1st result is :" + result + ", 2nd result is "+ result2);
    }
}

File: [0123][Hard][Best_Time_to_Buy_and_Sell_Stock_III]/Best_Time_to_Buy_and_Sell_Stock_III.py
# solution 1, dp general solution.
# leetcode time     cost : 156 ms
# leetcode memory   cost : 24.1 MB 
# Time  Complexity: O(N)
# Space Complexity: O(N)
import sys
class Solution:
    def maxProfit(self, prices):
        if not prices:
            return 0
        
        res = 0
        K_max = 2
        profit = [[ [0 for _ in range(2)] for _ in range(K_max+1)] for _ in range(len(prices))]
        
        profit[0][0][0], profit[0][0][1] = 0, -prices[0]
        profit[0][1][0], profit[0][1][1] = 0, -prices[0] 
        profit[0][2][0], profit[0][2][1] = 0, -prices[0]    

        # profit[i][k][j], i is the day, k is the transaction time, j is current holding status.
        for i in range(1,len(prices)):
            for k in range(1,K_max+1):
                profit[i][k][1] = max(profit[i-1][k][1], profit[i-1][k-1][0] - prices[i]) 
                profit[i][k][0] = max(profit[i-1][k][0], profit[i-1][k][1] + prices[i]) 
     
        end = len(prices) - 1
        res = profit[end][2][0]

        return res

def main():    
    nums1 = [1,3,5,0,0,3,1,4] #expect is 8
    obj = Solution()
    result = obj.maxProfit(nums1)
    print("return result is "+str(result));
    
    nums2 = [1,4,2] #expect is 3
    result = obj.maxProfit(nums2)
    print("return result is "+str(result));
    
    nums3 = [3,2,6,5,0,3] #expect is 7
    result = obj.maxProfit(nums3)
    print("return result is "+str(result));    
    
if __name__ =='__main__':
    main() 